Fast and Flexible GPU Quantization for Transformers.md

```diff
@@ -131,10 +131,72 @@ By zooming in (4 rows, 16 columns) we hopefully get a clearer picture of the lay
+
+In summary the decision is very much dependant on the accuracy/performance trade-off, as well as the specifics of the model architecture.
+
+|       Kernel       | Time (ms) | vs. FP16 |
+|:------------------:|:---------:|:--------:|
+| f16f16f16 (Torch)  |    600    |   1.0x   |
+| i8i8i32 (cuBLASLt) |    364    |   1.65x  |
+| i8i8i8 (cuBLASLt) |    308    |   1.95x  |
+
+#### FP16 output precision
+
+We previously touched upon the fact that I32 return type requires dequantizing outside of the matmul. Performance could be improved by fusing the dequant with the matmul and returning FP16 outputs. We can get this behaviour for free by using the GEMM `α` parameter to dequantize the outputs (the same way that we requantize for INT8 outputs), but this only works if we are applying **per-tensor** quantization, where the dequantize parameter is a single scalar {refer to per-scalar/per-channel section for more detail}.
+
+What if we require **per-channel** quantization i.e. using a vector to dequantize? In this scenario CUTLASS comes to the rescue by allowing the definition of a custom epilogue function, which is applied after the matrix multiplication, in a single fused kernel. For this scenario the GEMM + epilogue definition is expanded to 
+
+$$D=f_2(f_1(\alpha AB+\beta C, d)) $$
+The epilogue format comes from `EpilogueWithBroadcast` which applies a [binary operation](https://github.com/NVIDIA/cutlass/blob/master/include/cutlass/epilogue/thread/linear_combination_bias_elementwise.h#L215)`f1` between the matmul output and a column-wise broadcasted vector `d`, followed by an optional elementwise op `f2`.
+
+This might typically be a bias addition followed by an activation function (e.g. relu), but in our case we want `f1` to be a multiplication with the dequantize scalar. The epilogue is then plugged into `gemm::device::GemmUniversalWithBroadcast`.
+
+|       Kernel       | Time (ms) | vs. FP16 |
+|:------------------:|:---------:|:--------:|
+| f16f16f16 (Torch)  |    600    |   1.0x   |
+| i8i8i32 (CUTLASS) |    364    |   1.65x  |
+| i8i8f16 (CUTLASS) |    308    |   1.95x  |
+
+Whilst there might not be a huge improvement from FP16 output in terms of GEMM throughput, there are other peformance benefits:
+- 50% less data loaded in the next kernel (now FP16 instead of INT32)
+- Avoid fusion of the dequantize operator with the next kernel
+- Avoid having to load the dequantize vector in the next kernel (which CUTLASS pipelines the loading of)
```

Comment:
I'm not sure about the phrase "which CUTLASS pipelines the loading of". Not sure it reads quite right

Reply:
Had similar thoughts, any suggestions?